lib/AST/GenericEnvironment.cpp

```diff
@@ -40,24 +40,44 @@ GenericEnvironment::GenericEnvironment(
     addMapping(entry.first->castTo<GenericTypeParamType>(), entry.second);
 }
 
-void GenericEnvironment::setOwningDeclContext(DeclContext *newNowningDC) {
+/// Compute the depth of the \c DeclContext chain.
+static unsigned declContextDepth(const DeclContext *dc) {
+  unsigned depth = 0;
+  while (auto parentDC = dc->getParent()) {
+    ++depth;
```

Comment:
This is wrong for extensions of nested generic types. If you hit an extension dc you have to look at it's generic parameter list, and add up the number of outer parameter lists.